arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+//! more contexts.
+//!
+//! ## Permutation argument
+//!
+//! Communication between rows must be done using a permutation argument. The
+//! argument we use will be a generalisation of the one used in the [PlonK
+//! paper](https://eprint.iacr.org/2019/953).
+//!
+//! The construction of the permutations will be done using the methods prefixed
+//! `save` and `load`. The index of the current row and the index of the
+//! time the value has been written will be used to generate the permutation on
+//! the fly.
+//!
+//! The permutation argument described in the PlonK paper is a kind of "inverse
+//! lookup" protocol, like Plookup. The polynomials are defined as follows:
+//!
+//! ```text
+//!            Can be seen as T[f(X)] = Χ
+//!          --------
+//!          |      |
+//! f'(X) = f(X) + β X + γ
+//!                      |--- Can be seen as the evaluation point.
+//!                         |
+//!                         |
+//! g'(X) = g(X) + β σ(X) + γ
+//!          |      |
+//!          --------
+//!          Can be seen as T[g(X)] = σ(X)
+//! ```
+//!
+//! And from this, we build an accumulator, like for Plookup.
+//! The accumulator requires to coin two challenges, β and γ, and it must be
+//! done after the commitments to the columns have been absorbed.
+//! The verifier at the next step will verify that the challenges have been
+//! correctly computed.
+//! In the implementation, the accumulator will be computed after the challenges
+//! and the commitments. Note that the accumulator must also be aggregated, and
+//! the aggregation must be performed by the verifier at the next step.
+//!
+//! The methods `save` and `load` will accept as arguments only a column that is
+//! included in the permutation argument. For instance, `save_poseidon_state`
+//! will only accept columns with index 3, 4 and 5, where the
+//! `load_poseidon_state` will only accepts columns with index 0, 1 and 2.
+//!
+//! The permutations values will be saved in public values, and will contain the
+//! index of the row. The permutation values will be encoded with a 32 bits
+//! value (u32) as we can suppose a negligible probability that a user will use
+//! more than 2^32 rows.
+//!
+//! The permutation argument also generates constraints that will be
+//! homogeneized with the gadget constraints.
+//!
+//! Note all rows might require to use the permutation argument. Therefore, a
+//! selector will be added to activate/deactivate the permutation argument.
+//! When a method calls `save` or `load`, the selector will be activated. By
+//! default, the selector will be deactivated.
+//!
+//! TBD:
+//! - number of columns
+//! - accumulator column
+//! - folding of the permutation argument
+//!
+//! TBD/FIXME: do we use a additive permutation argument to increase the number
+//! of columns we can perform the permutation on?
+//!
+//! TBD/FIXME: We can have more than one permutation argument. For instance, we
```

Comment:
doing this prevents communication from column 1 to column 5
to reduce the degree we can use chunking : P = P_lo + X^k P_hi